host.py

```diff
@@ -34,29 +95,217 @@ def all_ports(self) -> list[dict[str, Any]]:
         return typing.cast(list[dict[str, Any]], r)
 
     @abstractmethod
-    def run(self, cmd: str, env: dict[str, str] = os.environ.copy()) -> Result:
+    def pretty_str(self) -> str:
         pass
 
+    @typing.overload
+    def run(
```

Comment:
Why is there 2 "run" functions with the same arguments but different return value types?

Reply:
Also there is no code for these functions

Reply:
it's for the `@typing.overload`.

https://docs.python.org/3/library/typing.html#overload

This describes certain combinations of signatures. In particular, this one exists so that mypy knows that `Host.run(..., text=False)` gives a binary `BinResult` and that `Host.run(..., text=True)` gives a string `Result`.

Reply:
It makes the code really hard to follow since it spans multiple lines and takes up a lot space in my opinion. This aspect of python typing makes the code bloated which I think python wanted to solve... ironically now it is even more bloated than C++.